python/opae.admin/opae/admin/tools/n5010tool.py

```diff
@@ -0,0 +1,182 @@
+            bit = N5010_PHY_ABSENT_0
+        elif p == 1:
+            offset = N5010_PHY_CSR_1
+            bit = N5010_PHY_ABSENT_1
+        elif p == 2:
+            offset = N5010_PHY_CSR_0
+            bit = N5010_PHY_ABSENT_0
+        elif p == 3:
+            offset = N5010_PHY_CSR_0
+            bit = N5010_PHY_ABSENT_1
+
+        val = regmap.reg_read(N5010_BMC_OFFSET + offset)
+        val &= bit
+        present = (val == 0)
+        qsfps.append(present)
+    return qsfps
+
+
+def pwr2str(qsfps, unit):
+    out = []
+    for rx_pwrs in qsfps:
+        out2 = []
+        for rx_pwr in rx_pwrs:
+            if rx_pwr == 0:
+                out3 = "   NA"
+            else:
+                # val read is in steps of 25.6 uW
+                val = 25.6 * rx_pwr
+                if unit == 'uW':
+                    out3 = f"{val:>5.0f}"
+                else:
+                    # conversion from uW to dBm
+                    val = 10.0 * math.log10(val * 0.001)
+                    out3 = f"{val:>5.2f}"
+            out2.append(out3)
+        out.append(out2)
+    return out
+
+def print_power(qsfps, uW, qspfs_insert):
+    """Print QSFP present and Rx pwr in chosen unit for each QSFP"""
+    if uW:
+        unit = 'uW'
+    else:
+        unit = 'dbm'
+    val = pwr2str(qsfps, unit)
+    port = 0
+    print("RX pwr: present\t lane1 \t lane2 \t lane3 \t lane4 \t unit")
+    for qsfp, present in zip(val, qspfs_insert):
+        if present:
+            qsfp_present = f"yes"
+        else:
+            qsfp_present = f"no"
+        p_str = f"Port {port}: {qsfp_present:>7s}\t {qsfp[0]:s}\t {qsfp[1]:s}\t {qsfp[2]:s}\t {qsfp[3]:s}\t {unit}"
+        print(p_str)
+        port += 1
+
+def main():
+    descr = 'Show QSFP status and optical Rx power levels for N501x card(s)'
+    parser = argparse.ArgumentParser(description=descr)
+    parser.add_argument('bdf', nargs='?',
+                        help=('PCIe address '
+                              '(eg 04:00.0 or 0000:04:00.0)'))
+    parser.add_argument('--uW', action='store_true',
+                        help='output in mW instead of dBm')
+
+    args = parser.parse_args()
+    if args.bdf:
+        args.bdf = normalize_bdf(args.bdf)
+        compatible = fpga.enum([{'pci_node.pci_address': str(args.bdf),
+                                 'pci_node.vendor_id': 0x1c2c}])
+    else:
+        compatible = fpga.enum([{'pci_node.vendor_id': 0x1c2c}])
+
+    if not compatible:
+        sys.stderr.write('No compatible devices found\n')
+        raise SystemExit(os.EX_USAGE)
+
+    for c in compatible:
+        print("\nQSFP info for card at: {}".format(c.pci_node.pci_address))
+        spi_bus = str(c.fme.spi_bus)
+
+        #Check BMC fw version >= 1.12.0
+        if not (c.fme.bmcfw_version >= max10_or_nios_version(0x10c00)):
+            sys.stderr.write(f'Device with BMC {c.fme.bmcfw_version} not supported\n')
+        else:
+            regmap = init_regmap(regmap_path+spi_bus)
```

Comment:
add this code in try cath

Reply:
Sorry for asking: What do you want to achieve with the try block? init_regmap() raise exception if no root privileges or no debug fs. Typically it would also fail for next card, so there is no need to continue. If you want get passed an error card you can provide bdf argument.